### ISSUE TITLE
test-serialization failure/warnings due to timezone

### ISSUE THREAD

**Issue by nealrichardson** [sparklyr/sparklyr#2543]
I observed a test failure and some warnings locally. Here's output (I dropped some `print()`s in to see what was comparing wrong, as well as sessionInfo):

```
> devtools::test(filter="^serialization$")
Loading sparklyr

Attaching package: ‘dplyr’

The following object is masked from ‘package:testthat’:

    matches

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

Testing sparklyr
✔ |  OK F W S | Context
⠸ |  14       | serialization# A tibble: 4 x 2
  date       name           
  <date>     <chr>          
1 1961-01-19 John F. Kennedy
2 1969-12-31 NA             
3 1981-01-19 Ronald Reagan  
4 2001-01-19 George W. Bush 
⠴ |  25 1     | serialization[[1]]
[1] "2018-01-01 UTC"

[[2]]
[1] "2018-01-01 UTC"

[[3]]
[1] "2018-01-01 08:00:00 UTC"

[[4]]
[1] "2018-01-01 11:33:00 UTC"

[[5]]
[1] "2018-01-01 02:33:00 UTC"

✖ |  35 1 4   | serialization [44.8 s]
────────────────────────────────────────────────────────────────────────────────
test-serialization.R:285: failure: collect() can retrieve date types successfully
as.list(df) not equal to as.list(...).
Component "date": Mean relative difference: 0.0002145117

test-serialization.R:332: warning: collect() can retrieve as.POSIXct fields with timezones
'tzone' attributes are inconsistent

test-serialization.R:332: warning: collect() can retrieve as.POSIXct fields with timezones
'tzone' attributes are inconsistent

test-serialization.R:332: warning: collect() can retrieve as.POSIXct fields with timezones
'tzone' attributes are inconsistent

test-serialization.R:332: warning: collect() can retrieve as.POSIXct fields with timezones
'tzone' attributes are inconsistent
────────────────────────────────────────────────────────────────────────────────

══ Results ═════════════════════════════════════════════════════════════════════
Duration: 44.8 s

OK:       35
Failed:   1
Warnings: 4
Skipped:  0

Frustration is a natural part of programming :)
```

sessionInfo:

```
R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_0.8.5         sparklyr_1.2.0.9001 testthat_2.3.2     

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.3        dbplyr_1.4.3      compiler_3.6.2    pillar_1.4.4     
 [5] r2d3_0.2.3        prettyunits_1.1.1 base64enc_0.1-3   remotes_2.1.0    
 [9] tools_3.6.2       uuid_0.1-2        digest_0.6.25     pkgbuild_1.0.6   
[13] pkgload_1.0.2     jsonlite_1.6      memoise_1.1.0     tibble_3.0.1     
[17] lifecycle_0.2.0   pkgconfig_2.0.3   rlang_0.4.6       cli_2.0.2        
[21] DBI_1.1.0         rstudioapi_0.11   yaml_2.2.0        xml2_1.3.2       
[25] httr_1.4.1        withr_2.1.2       rappdirs_0.3.1    htmlwidgets_1.5.1
[29] askpass_1.1       globals_0.12.5    generics_0.0.2    desc_1.2.0       
[33] fs_1.3.1          vctrs_0.3.0       devtools_2.2.1    tidyselect_1.1.0 
[37] rprojroot_1.3-2   forge_0.2.0       glue_1.4.0        R6_2.4.1         
[41] processx_3.4.2    fansi_0.4.1       sessioninfo_1.1.1 tidyr_1.0.2      
[45] purrr_0.3.4       callr_3.4.3       magrittr_1.5      htmltools_0.4.0  
[49] codetools_0.2-16  backports_1.1.6   ps_1.3.2          ellipsis_0.3.0   
[53] usethis_1.6.1     assertthat_0.2.1  config_0.3        openssl_1.4.1    
[57] rjson_0.2.20      crayon_1.3.4     
```

This is also with spark 3.0.0.preview2.

It would seem that something somewhere is coercing everything to UTC, and Date types to something that would be timezone-aware, so they get shifted if you're not in UTC, which my laptop is not. 

You wouldn't see the test failure on CI because those machines are on UTC, and I would guess that the warnings are happening but don't fail the build.

**Comment by nealrichardson** [comment]
Note that if you enable `arrow`, the Date issue is resolved (the other test is currently skipped; I'm in this file because I'm trying to see if we can unskip it).

Probably worth considering the issues observed in #2439 and #2173 here too.

**Comment by yl790** [comment]
Hey @nealrichardson I think I know why this is happening... When `Date` type got shipped back from Spark to R without arrow, someone decided to use `java.sql.Date.toString` to convert the date into its string form :astonished:  , and that's where it gets an offset from the JVM timezone. I'll fix this ASAP.

**Comment by yl790** [comment]
and also there is somewhere else that introduced some floating point rounding error :facepalm: which explains things like "Mean relative difference: 0.0002145117"

**Comment by yl790** [comment]
Update: was aiming to get this fixed today but the new dplyr 1.0 not working well with sparklyr complicated the matter quite a bit... Might need to wait and see whether it was a dplyr problem or not

**Comment by yl790** [comment]
We now set a non-UTC timezone in sparklyr's github CI workflow to uncover portability issues like this one.
